lib/pbench/server/api/resources/query_apis/__init__.py

```diff
@@ -820,61 +818,82 @@ def _bulk_dispatch(
+                )
+            else:
+                report = BulkResults(errors=0, count=0, report={})
 
-        # If we don't have an Elasticsearch index map, then the dataset isn't
-        # indexed and we skip the Elasticsearch actions.
-        if IndexMap.exists(dataset):
-            # Build an Elasticsearch instance to manage the bulk update
-            elastic = Elasticsearch(self.elastic_uri)
-            map = IndexMap.stream(dataset=dataset)
+            summary: JSONOBJECT = {
+                "ok": report.count - report.errors,
+                "failure": report.errors,
+            }
+            attributes: JSONOBJECT = {"summary": summary}
+            auditing["attributes"] = attributes
 
-            # NOTE: because both generate_actions and streaming_bulk return
-            # generators, the entire sequence is inside a single try block.
+            # Let the subclass complete the operation
             try:
-                results = helpers.streaming_bulk(
-                    elastic,
-                    self.generate_actions(dataset, context, map),
-                    raise_on_exception=False,
-                    raise_on_error=False,
-                )
-                report = self._analyze_bulk(results, context)
+                self.complete(dataset, context, summary)
             except Exception as e:
-                raise APIInternalError("Unexpected backend error '{e}'") from e
-        elif context["attributes"].require_map and self.expect_index(dataset):
-            # If the dataset has no index map, the bulk operation requires one,
-            # and we expect one to appear, fail rather than risking abandoning
-            # Elasticsearch documents.
-            raise APIAbort(
-                HTTPStatus.CONFLICT,
-                f"Operation unavailable: dataset {dataset.resource_id} is not indexed.",
-            )
-        else:
-            report = BulkResults(errors=0, count=0, report={})
+                raise APIInternalError(f"Unexpected completion error '{e}'") from e
+        except Exception as e:
+            # Make sure that we correctly audit errors, and clean up the
+            # operational state.
+            auditing["attributes"] = {"message": str(e)}
+            auditing["status"] = AuditStatus.FAILURE
+            auditing["reason"] = AuditReason.INTERNAL
 
-        summary: JSONOBJECT = {
-            "ok": report.count - report.errors,
-            "failure": report.errors,
-        }
-        auditing: dict[str, Any] = context["auditing"]
-        attributes: JSONOBJECT = {"summary": summary}
-        auditing["attributes"] = attributes
+            # The DELETE operator removes the "sync" context to signal that
+            # the operations table rows no longer exist, so check.
+            if context.get("sync"):
+                context["sync"].error(dataset, f"update error {e}")
```

Comment:
Yeah, the latter is what I originally did. It was a *bug*.

`DELETE` is a weird case as the object goes away on successful completion. I had to devise a signal to avoid referencing the row ID on return from `complete`; and the convention I'd implemented (which is admittedly a bit suble) was `del context["sync"]` ... re-accessing that here instead of using `sync` preserves that convention instead of making more changes. And I wrote a comment trying to clarify that, but apparently it wasn't clear enough ...